zomes/wiki/code/src/page.rs

```diff
@@ -95,75 +84,95 @@ pub fn page_def() -> ValidatingEntryType {
-            let page_anchor = utils::anchor("wiki_pages", "all_pages")?;
-            hdk::utils::commit_and_link(
-                &Page::initial(title.clone()).entry(),
-                &page_anchor,
-                "anchor->Page",
-                &title,
-            )
-        }
-        Some(_) => Ok(address),
+pub fn create_page_if_non_existent(
+    title: String,
+    sections: Vec<Address>,
+    timestamp: String,
+) -> ZomeApiResult<Address> {
+    let anchor_address = holochain_anchors::anchor("wiki_pages".to_string(), title.clone())?;
+    if let Some(address) = hdk::get_links(
+        &anchor_address,
+        LinkMatch::Exactly("anchor->Page"),
+        LinkMatch::Any,
+    )?
+    .addresses()
+    .pop()
+    {
+        Ok(address)
+    } else {
+        let page_entry = Page::from(title.clone(), sections, timestamp).entry();
+
+        hdk::utils::commit_and_link(&page_entry, &anchor_address, "anchor->Page", &title)
     }
 }
-
 pub fn create_page_with_sections(
-    sections: Vec<Section2>,
+    sections: Vec<Section>,
     title: String,
-    timestamp: u64,
+    timestamp: String,
 ) -> ZomeApiResult<String> {
-    let page_address = create_page_if_non_existent(title.clone())?;
-    let sections: Vec<Address> = sections
-        .into_iter()
-        .map(|element| {
-            hdk::api::commit_entry(&Section::from(element, page_address.clone()).entry())
-        })
-        .filter_map(Result::ok)
-        .collect();
+    let anchor_address = holochain_anchors::anchor("wiki_pages".to_string(), title.clone())?;
 
-    hdk::api::update_entry(
-        Page::from(title.clone(), sections, Some(timestamp)).entry(),
-        &page_address,
+    create_page_if_non_existent(
+        title.clone(),
+        sections
+            .into_iter()
+            .map(|section| hdk::api::commit_entry(&section.from(anchor_address.clone()).entry()))
+            .filter_map(Result::ok)
+            .collect(),
+        timestamp,
     )?;
 
     Ok(title)
 }
-pub fn update_page(sections: Vec<Address>, title: String, timestamp: u64) -> ZomeApiResult<String> {
-    let page_address = create_page_if_non_existent(title.clone())?;
+pub fn update_page(
+    sections: Vec<Address>,
+    title: String,
+    timestamp: String,
+) -> ZomeApiResult<String> {
+    let page_address =
+        create_page_if_non_existent(title.clone(), sections.clone(), timestamp.clone())?;
     hdk::api::update_entry(
-        Page::from(title.clone(), sections, Some(timestamp)).entry(),
+        Page::from(title.clone(), sections, timestamp).entry(),
         &page_address,
     )?;
     Ok(title)
 }
-pub fn get_page(title: String) -> ZomeApiResult<JsonString> {
-    utils::get_entry(Page::initial(title).entry().address())
-}
-
-pub fn get_titles() -> ZomeApiResult<Vec<String>> {
-    let anchor_address = utils::anchor("wiki_pages", "all_pages")?;
+pub fn get_page(title: String) -> ZomeApiResult<Page> {
     Ok(hdk::utils::get_links_and_load_type::<Page>(
```

Comment:
Change this not to use get_links_and_load_type, because this makes unnecessary calls. Split it into get_links, get the first link and do a get_entry.